src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -236,6 +236,34 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let result = Scalar::from_uint(truncated_bits, layout.size);
                 self.write_scalar(result, dest)?;
             }
+
+            "ptr_offset_from" => {
+                let a = self.read_immediate(args[0])?.to_scalar()?.to_ptr()?;
+                let b = self.read_immediate(args[1])?.to_scalar()?.to_ptr()?;
+                if a.alloc_id != b.alloc_id {
+                    throw_ub_format!(
+                        "ptr_offset_from cannot compute offset of pointers into different \
+                        allocations.",
+                    );
+                }
+                let usize_layout = self.layout_of(self.tcx.types.usize)?;
+                let a_offset = ImmTy::from_uint(a.offset.bytes(), usize_layout);
+                let b_offset = ImmTy::from_uint(b.offset.bytes(), usize_layout);
+                let (val, overflowed, _) = self.overflowing_binary_op(
+                    BinOp::Sub, a_offset, b_offset,
+                )?;
+                if overflowed {
+                    throw_ub_format!(
+                        "second argument to `ptr_offset_from` must be smaller than first",
```

Comment:
Err, is this actually a requirement? This doesn't seem to be in the `offset_from()` documentation, and given that it returns `isize` rather than `usize`, I would not expect such a limitation to exist.

(If this were a requirement, we could also use `sub nuw` on the LLVM side, but there are comments explicitly saying that we can't.)

Reply:
Right. This is a point of discussion around `offset_from`. I initially thought this was a requirement and implemented it thus. I still think it should be a requirement, but that's up to T-libs I guess.

Reply:
Looking at the [documentation](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.offset_from), currently it does not seem to be a requirement, so it seems o be prudent to implement it as such?